storage/range_data_iter.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Spencer Kimball (spencer.kimball@gmail.com)
+
+package storage
+
+import (
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util/encoding"
+)
+
+// keyRange is a helper struct for the rangeDataIterator.
+type keyRange struct {
+	start, end proto.EncodedKey
+}
+
+// rangeDataIterator provides a complete iteration over all key / value
+// rows in a range, including all system-local metadata and user data.
+// The ranges keyRange slice specifies the key ranges which comprise
+// all of the range's data.
+//
+// A rangeDataIterator provides the same API as an Engine iterator
+// with the exception of the Seek() method.
+type rangeDataIterator struct {
+	curIndex int
+	ranges   []keyRange
+	iter     engine.Iterator
+}
+
+func newRangeDataIterator(r *Range, e engine.Engine) *rangeDataIterator {
+	startKey := r.Desc.StartKey
+	if startKey.Equal(engine.KeyMin) {
+		startKey = engine.KeyLocalMax
+	}
+	ri := &rangeDataIterator{
+		ranges: []keyRange{
+			{
+				start: engine.MVCCEncodeKey(responseCacheKeyPrefix(r.Desc.RaftID)),
```

Comment:
It seems error-prone to keep this list of keys up to date. Can we move to a common scheme for range-related data, e.g. to encode RaftID consistently at the start of the key so we can iterate over everything beginning with that prefix? We might need more than one scheme (e.g. one based on RaftID and one based on StartKey), but that would be better than 6 (and growing - I'll need to add the raft-related keys from my branch).